jina/serve/networking.py

```diff
@@ -145,13 +146,105 @@ class GrpcConnectionPool:
     K8S_PORT_USES_BEFORE = 8081
     K8S_PORT = 8080
 
+    class _ConnectionStubs:
+        """
+        Maintains a list of grpc stubs available for a particular connection
+        """
+
+        STUB_MAPPING = {
+            'jina.JinaControlRequestRPC': jina_pb2_grpc.JinaControlRequestRPCStub,
+            'jina.JinaDataRequestRPC': jina_pb2_grpc.JinaDataRequestRPCStub,
+            'jina.JinaSingleDataRequestRPC': jina_pb2_grpc.JinaSingleDataRequestRPCStub,
+            'jina.JinaRPC': jina_pb2_grpc.JinaRPCStub,
+        }
+
+        def __init__(self, address, channel, summary):
+            self.address = address
+            self.channel = channel
+            self._summary_time = summary
+            self._initialized = False
+
+        # This has to be done lazily, because the target endpoint may not be available
+        # when a connection is added
+        def _init_stubs(self):
+            available_services = GrpcConnectionPool.get_available_services(self.address)
+            stubs = defaultdict(lambda: None)
+            for service in available_services:
+                # skip the reflection service, it will not be used anymore
+                if service != 'grpc.reflection.v1alpha.ServerReflection':
+                    stubs[service] = self.STUB_MAPPING[service](self.channel)
+            self.control_stub = stubs['jina.JinaControlRequestRPC']
+            self.data_list_stub = stubs['jina.JinaDataRequestRPC']
+            self.single_data_stub = stubs['jina.JinaSingleDataRequestRPC']
+            self.stream_stub = stubs['jina.JinaRPC']
+            self._initialized = True
+
+        async def send_requests(
+            self, requests: List[Request], metadata, compression
+        ) -> Tuple:
+            if not self._initialized:
```

Comment:
can't we avoid this and simply initialize at `__init__` time? I prefer a more stateless approach

Reply:
unfortunately not. The lazy init here is connecting to the server's grpc reflection endpoint, but the __init__ is called on adding connections. At this time the server might not be ready yet. This is why I am delaying this until sending of the first data request. Then the server should be ready (or the request would fail anyway).